### PR TITLE
[flang][OpenMP] Fix privatization when critical is present

### LINKED ISSUE #75767: [Flang][OpenMP] Incorrect execution result of a variable in critical construct in parallel construct with default(firstprivate) clause
```
Version of flang-new : 18.0.0(76bbbcb41bcf4a1d7a26bb11b78cf97b60ea7d4b)
```

The execution result of a variable in `critical` construct in `parallel` construct with `default(firstprivate)` clause is incorrect.  
The execution result is correct if:
- remove `critical` construct, or
- change `default(firstprivate)` to `firstprivate(j)`

The following are the test program, Flang-new, Gfortran and ifort compilation/execution result.

snf_ng_01_1.f90:
```fortran
program main
  INTEGER :: j
  j=2
!$OMP PARALLEL default(firstprivate)
!$OMP critical
  j=200
!$OMP END critical
!$OMP END PARALLEL
  if (j.ne.2) print *,'err var j=',j
  print *,'pass'
end program main
```

```
$ export OMP_NUM_THREADS=2; flang-new -fopenmp snf_ng_01_1.f90; ./a.out
 err var j= 200
 pass
$
```

```
$ export OMP_NUM_THREADS=2; gfortran -fopenmp snf_ng_01_1.f90; ./a.out
 pass
$
```

```
$ export OMP_NUM_THREADS=2; ifort -qopenmp snf_ng_01_1.f90; ./a.out
 pass
$
```

### PR DESCRIPTION
**Pull request by luporl**
When a critical construct is present inside another construct where
privatizations may occur, such as a parallel construct, some
privatizations are skipped if the corresponding symbols are defined
inside the critical section only (see the example below).

This happens because, while critical constructs have a "body", they
don't have a separate scope (which makes sense, since no
privatizations can occur in them). Because of this, in semantics
phase, it's not possible to insert a new host association symbol,
but instead the symbol from the enclosing context is used directly.

This makes symbol collection in DataSharingProcessor consider the
new symbol to be defined by the critical construct, instead of by
the enclosing one, which causes the privatization to be skipped.

Example:
```
!$omp parallel default(firstprivate)
  !$omp critical
     i = 200
  !$omp end critical
!$omp end parallel
```

This patch fixes this by identifying constructs where
privatizations may not happen and skipping them during the
collection of nested symbols. Currently, this seems to happen only
with critical constructs, but others can be easily added to the
skip list, if needed.

Fixes https://github.com/llvm/llvm-project/issues/75767

### CODE REVIEW

File: flang/test/Lower/OpenMP/critical.f90 [hunk 1]
```diff
@@ -51,3 +51,27 @@ subroutine predetermined_privatization()
   end do
   !$omp end parallel do
 end
+
+! https://github.com/llvm/llvm-project/issues/75767
+!CHECK-LABEL: func @_QPnested_privatization(
+subroutine nested_privatization()
```
Comment (luporl) [inline flang/test/Lower/OpenMP/critical.f90:57]: Looking again, `nested_privatization` doesn't seem to properly describe the scenario where the issue happened.
Maybe `parallel_critical_privatization` would be a better name?

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 2]
```diff
@@ -262,6 +262,15 @@ void DataSharingProcessor::insertLastPrivateCompare(mlir::Operation *op) {
   }
 }
 
+static bool mayHavePrivatizations(const lower::pft::Evaluation &eval) {
+  if (const parser::OpenMPConstruct *construct =
```
Comment (luporl) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:266]:
```suggestion
  if (const auto *construct =
```

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp [hunk 3]
```diff
@@ -262,6 +262,15 @@ void DataSharingProcessor::insertLastPrivateCompare(mlir::Operation *op) {
   }
 }
 
+static bool mayHavePrivatizations(const lower::pft::Evaluation &eval) {
```
Comment (ergawy) [inline flang/lib/Lower/OpenMP/DataSharingProcessor.cpp:265]: Maybe we should instead update the `OMPConstructSymbolVisitor` as such:
```diff
diff --git a/flang/lib/Lower/OpenMP/DataSharingProcessor.h b/flang/lib/Lower/OpenMP/DataSharingProcessor
.h
index 80a956de35ba..d1f440f841c0 100644
--- a/flang/lib/Lower/OpenMP/DataSharingProcessor.h
+++ b/flang/lib/Lower/OpenMP/DataSharingProcessor.h
@@ -44,7 +44,9 @@ private:
     void Post(const T &) {}
 
     bool Pre(const parser::OpenMPConstruct &omp) {
-      currentConstruct = &omp;
+      if (!std::holds_alternative<parser::OpenMPCriticalConstruct>(omp.u))
+        currentConstruct = &omp;
+
       return true;
     }
```

This way we keep the active construct tracking logic in one place (i.e. the visitor) and don't have to add a new function or a new check in `collectSymbolsInNestedRegions(...)`.

**Review by ergawy — COMMENTED** [review]
Thanks for the fix @luporl! I have a different proposal toachive the same result. Let me know what you think!

File: flang/lib/Lower/OpenMP/DataSharingProcessor.cpp (hunk 3 above)
Comment (luporl) [reply flang/lib/Lower/OpenMP/DataSharingProcessor.cpp]: Thanks for the review! Your suggestion worked well, so I've switched to it in the last commit.

**Review by ergawy — APPROVED** [review]

**Pull request merged**